### PR TITLE
added vmware_add_nic module

### PR DESCRIPTION
**Pull request by mholiv**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME
vmware_add_nic

##### ANSIBLE VERSION

```
ansible --version
ansible 2.2.0.0
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY

A module that allows the user to add nics to virtual hosts to VMware

### CODE REVIEW

**Comment by mholiv** [comment]
All checks now passed yay!

**Comment by dagwieers** [comment]
The module name should be something like **vmware_guest_nic** as it manages the nic (not just add them, that would not be idempotent). The other concern I have is that this functionality is currently taken care of by vmware_guest (or at least there is overlapping functionality.

My preference would be that any change to the vmware_guest call for a given VM would automatically modify that VM to the requested form. So in this case hot-add/hot-remove NICs. Not sure if that is currently possible or wanted.

**Comment by jctanner** [comment]
@mholiv is this something you still want to work on?

needs_info

**Comment by dagwieers** [comment]
+label new_module

**Comment by cdelgehier** [comment]
ahah, VMWare talk to me in french :s But it doesn't work 

```python
>>> vm = content.searchIndex.FindByUuid(None, '4218eab0-a5a9-b98a-8d42-ec846d22fe0b', True)
>>> for device in vm.config.hardware.device:
...     device.deviceInfo.label
... 
'IDE 0'
'IDE 1'
'Contrôleur PS2 0'
'Contrôleur PCI 0'
'Contrôleur SIO 0'
'Clavier '
'Périphérique de pointage'
'Carte vidéo '
'Périphérique VMCI'
'Contrôleur SCSI 0'
'Disque dur 1'
'Adaptateur réseau 1'
```

```
ok: [testvm -> localhost] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "datacenter": "Blue", 
            "dvs": false, 
            "hostname": "172.16.17.2", 
            "label": "Adaptateur réseau 1", 
            "network_name": null, 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "state": "absent", 
            "type": "vmxnet3", 
            "username": "vmware", 
            "validate_certs": false, 
            "vm_path": "/DEV/TEST/testvm"
        }
    }, 
    "msg": "No NIC with the name: Adaptateur réseau 1 was found"
}
```

File: lib/ansible/modules/cloud/vmware/vmware_add_nic.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = '''
+---
+module: vmware_nic
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_add_nic.py:20]: Could you please match the name of module - `vmware_add_nic` ?

File: lib/ansible/modules/cloud/vmware/vmware_add_nic.py [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+short_description: Create, Update, or Remove a NIC from a VM
+description:
+     - Allows non-idempotent creation of VMs. Updates and removals are idempotent.
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_add_nic.py:23]: Do you mean `Allows non-idempotent creation of NIC for VMs` ?

File: lib/ansible/modules/cloud/vmware/vmware_add_nic.py [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+options:
+  vm_path:
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_add_nic.py:25]: Could you please change this to `folder` to match other VMware modules?

File: lib/ansible/modules/cloud/vmware/vmware_add_nic.py [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+            if new_obj.name != 'vm':
+                # 'vm' is an invisible folder that exists at the datacenter root so we ignore it
+                path_elements.append(new_obj.name)
+
+            objwalk(new_obj, path_elements)
+
+    return path_elements
+
+
+def get_vm_object(module, conn, path, datacenter):
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_add_nic.py:131]: Could you please use `find_vm_by_id` from vmware module_utils ?

File: lib/ansible/modules/cloud/vmware/vmware_add_nic.py [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+def create_nic(module, conn, vm, desired_nic):
+    vm_spec = vim.vm.ConfigSpec()
+    nic_spec = vim.vm.device.VirtualDeviceSpec()
+    changes = []
+
+    nic_spec.operation = vim.vm.device.VirtualDeviceSpec.Operation.add
+    nic_spec.device = getattr(vim.vm.device, desired_nic['type'])()
+
+    if desired_nic['dvs']:
+        pg_obj = get_obj_by_name(conn, [vim.dvs.DistributedVirtualPortgroup], desired_nic['network'])
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_add_nic.py:185]: Could you please use `find_dvs_by_name` from vmware module_utils ?

File: lib/ansible/modules/cloud/vmware/vmware_add_nic.py [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+
+    return True
+
+
+def main():
+
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        dict(
+                vm_path=dict(required=True, type='str'),
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_add_nic.py:350]: Can you please include `uuid` as parameter for finding VM ? you can use `find_by_vm_id` API for finding VM using uuid just like other VMware modules do.

File: lib/ansible/modules/cloud/vmware/vmware_add_nic.py [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+            changed=True
+            update_nic(module, conn, proper_vm, desired_nic, all_nics)
+        else:
+            changed=False
+
+        module.exit_json(changed=changed)
+
+
+
+from ansible.module_utils.basic import *
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_add_nic.py:429]: Please use particular import from these modules.

File: lib/ansible/modules/cloud/vmware/vmware_add_nic.py [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+    elif state == 'update':
+        if needs_update(module, desired_nic, all_nics):
+            changed=True
+            update_nic(module, conn, proper_vm, desired_nic, all_nics)
+        else:
+            changed=False
+
+        module.exit_json(changed=changed)
+
+
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_add_nic.py:427]: Could you please add integration tests for this module ? Please take a look at various VMware related tests in `test/integration` directory.

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by cdelgehier** [comment]
Adding a card to a vswitch works.
But when you pass the `dvs` to `true` for an existing DVSwitch, we get an error
```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_m6fqrU/ansible_module_vmware_nic.py", line 433, in <module>
    main()
  File "/tmp/ansible_m6fqrU/ansible_module_vmware_nic.py", line 402, in main
    results = create_nic(module, conn, proper_vm, desired_nic)
  File "/tmp/ansible_m6fqrU/ansible_module_vmware_nic.py", line 202, in create_nic
    nic_spec.device.backing.deviceName = desired_nic['network']
  File "/usr/local/lib/python2.7/dist-packages/pyVmomi/VmomiSupport.py", line 537, in __setattr__
    CheckField(self._GetPropertyInfo(name), val)
  File "/usr/local/lib/python2.7/dist-packages/pyVmomi/VmomiSupport.py", line 342, in GetPropertyInfo
    raise AttributeError(name)
AttributeError: deviceName
```

Otherwise I have to bypass the language problem by changing the conditions:

```if nic['nic_obj'].deviceInfo.label == desired_nic['label']:```
to 
```if nic['nic_obj'].deviceInfo.label.rsplit(None, 1)[-1] == desired_nic['label']:```

So the `label` parameter changes from `Network adapter 1` to `1`

**Comment by dagwieers** [comment]
So the module freeze deadline for v2.4 is getting closer (2017-08-29). Now is the time to finish up, get it reviewed and merged with no delay.

**Comment by Akasurde** [comment]
@mholiv Are you still working on this ?

**Comment by mholiv** [comment]
No. This should be encompassed in the general VMware module set anyways.

**Comment by Akasurde** [comment]
Closing this due to inactivity of contributor.

**Pull request closed without merging**